testsuite/tests/typing-extensions/floatarray.ml

```diff
@@ -0,0 +1,85 @@
+(* TEST
+   * expect
+*)
+
+let x = [|1.;2.|]
+;;
+[%%expect {|
+val x : float array = [|1.; 2.|]
+|}]
+
+let x : floatarray = [|1.; 2.|]
+;;
+[%%expect {|
+val x : floatarray = [|1.; 2.|]
+|}]
+
+let x = ([|1.; 2.|] : floatarray)
+;;
+[%%expect {|
+val x : floatarray = [|1.; 2.|]
+|}]
+
+let f (a : floatarray) = match a with [|x|] -> x | _ -> assert false
+;;
+[%%expect {|
+val f : floatarray -> float = <fun>
+|}]
+
+let _ = f [|1.|]
+;;
+[%%expect {|
+- : float = 1.
+|}]
+
+(* Does not work without the annotation *)
+
+let f a = match a with [|_|] -> Float.Array.length a | _ -> assert false
+;;
+[%%expect {|
+Line 1, characters 51-52:
+1 | let f a = match a with [|_|] -> Float.Array.length a | _ -> assert false
+                                                       ^
+Error: This expression has type 'a array
+       but an expression was expected of type Float.Array.t = floatarray
+|}]
+
+type s = floatarray
+type t = s
+let x : t = [||]
+;;
+[%%expect {|
+type s = floatarray
+type t = s
+val x : t = [||]
+|}]
+
+let f a =
+  let _ = Float.Array.length a in
+  match a with
+  | [||] -> ()
+  | _ -> ()
+;;
+[%%expect{|
+val f : Float.Array.t -> unit = <fun>
+|}, Principal{|
+Line 4, characters 4-8:
+4 |   | [||] -> ()
+        ^^^^
+Warning 18 [not-principal]: this type-based field disambiguation is not principal.
+val f : Float.Array.t -> unit = <fun>
```

Comment:
The warning here is a bit odd, given that this typing is in fact principal (there is no other valid type than `floatarray` for this pattern), and that we do *not* have a warning in the actually-not-principal case `let x = [| 1.2 |]` (no type annotations, `array` chosen by default).

Reply:
The warning is the same as we get with record disambiguation. It's a bit of a stretch, but I think of it as being not principal between the two options of "types as `floatarray`" and "gives a type error".

Reply:
Note: also "field disambiguation" is wrong in the error message, maybe the functor would need to be made more parametric on the sort of disambiguation.

@lpw25 the difference I make with variant/record disambiguation is that, for variants, you also get a warning when you construct variants in a non-principal way, whereas here the warning shows up in cases that are actually principal, and there is no warning in cases that are not.

Reply:
> for variants, you also get a warning when you construct variants in a non-principal way

I don't think that you do:
```ocaml
# type t = Foo;;
type t = Foo
# type s = Foo;;
type s = Foo
# let x = Foo;;
val x : s = Foo
```
at least not by default. I don't think anyone actually turns on warning 41.

Reply:
Also worth noting:
```
# module M = struct
    type t = Foo
  end;;
    module M : sig type t = Foo end
# type s = Foo;;
type s = Foo
# #warnings "+41";;
# let x = Foo;;
val x : s = Foo
# let x : M.t = Foo;;
val x : M.t = M.Foo
```

Reply:
Erm, you are correct. So the behavior in this PR is as nonsensical as the behavior already implemented for variants. I'll drop my criticism of this particular testsuite example, but I remain of the opinion that we should get a non-principality warning when we write non-principal code. (And I think that doing this properly in this PR may require a code modification.)